src/main/scala/uk/sky/fs2/kafka/topicloader/TopicLoader.scala

```diff
@@ -148,23 +157,34 @@ trait TopicLoader {
       partitionInfo <- topics.toList.flatTraverse(consumer.partitionsFor)
     } yield partitionInfo.map(pi => TopicPartition(pi.topic, pi.partition)).toSet
 
-  private def emitRecordRemovingConsumedPartition[F[_] : Monad : LoggerFactory, K, V](
+  private def emitRecordRemovingConsumedPartition[K, V](
       t: HighestOffsetsWithRecord[K, V],
       r: ConsumerRecord[K, V]
-  ): F[HighestOffsetsWithRecord[K, V]] = {
-    val logger = LoggerFactory[F].getLogger
-
+  ): HighestOffsetsWithRecord[K, V] = {
     val partitionHighest: Option[Long] = t.partitionOffsets.get(TopicPartition(r.topic, r.partition))
 
-    val reachedHighest: OptionT[F, TopicPartition] = for {
-      offset  <- OptionT.fromOption[F](partitionHighest)
-      highest <- OptionT.fromOption[F](if (r.offset >= offset) TopicPartition(r.topic, r.partition).some else None)
-      _       <- OptionT.liftF(logger.warn(s"Finished loading data from ${r.show} at offset ${r.offset}"))
+    val reachedHighest: Option[TopicPartition] = for {
+      offset  <- partitionHighest
+      highest <- Option.when(r.offset >= offset)(TopicPartition(r.topic, r.partition))
     } yield highest
 
-    val updatedHighests = reachedHighest.fold(t.partitionOffsets)(highest => t.partitionOffsets - highest)
     val emittableRecord = partitionHighest.collect { case h if r.offset <= h => r }
-    updatedHighests.map(HighestOffsetsWithRecord(_, emittableRecord))
+
+    reachedHighest match {
+      case Some(highest) =>
+        HighestOffsetsWithRecord(
+          partitionOffsets = t.partitionOffsets - highest,
+          consumerRecord = emittableRecord,
+          partitionLast = PartitionLast(highest, r.offset).some
+        )
```

Comment:
If we've reached the highest offset for this partition, store the Topic Partition and Offset inside the `HighestOffsetsWithRecord` so we can signal this upstream